spark/v4.0/spark/src/main/java/org/apache/iceberg/spark/security/IcebergHiveConnectorDelegationTokenProvider.java

```diff
@@ -0,0 +1,240 @@
+    String metastoreUri = sparkConf.get(SPARK_SQL_CATALOG_PREFIX + catalogName + URI_KEY, "");
+    String principal = sparkConf.get(SPARK_SQL_CATALOG_PREFIX + catalogName + PRINCIPAL_KEY, "");
+    boolean isHiveType =
+        CATALOG_TYPE.equalsIgnoreCase(
+            sparkConf.get(SPARK_SQL_CATALOG_PREFIX + catalogName + TYPE_KEY, ""));
+    if (metastoreUri.isEmpty()
+        || principal.isEmpty()
+        || !isHiveType
+        || !UserGroupInformation.isSecurityEnabled()) {
+      return false;
+    }
+
+    try {
+      Credentials credentials = UserGroupInformation.getCurrentUser().getCredentials();
+      Token<?> currentToken = credentials.getToken(new Text(metastoreUri));
+      return currentToken == null;
+    } catch (IOException ex) {
+      LOG.error(
+          "Failed to get current user credentials for catalog {}: {}",
+          catalogName,
+          ex.getMessage(),
+          ex);
+      throw new RuntimeException(ex);
+    }
+  }
+
+  IMetaStoreClient createHmsClient(HiveConf conf) throws Exception {
+    return new HiveMetaStoreClient(conf, null, false);
+  }
+
+  @Override
+  public Option<Object> obtainDelegationTokens(
+      Configuration hadoopConf, SparkConf sparkConf, Credentials creds) {
+    Map<String, IMetaStoreClient> hmsClientCache = Maps.newHashMap();
+    Set<String> requireTokenCatalogs = getRequireTokenCatalogs(sparkConf);
+    LOG.debug("Require token Hive catalogs: {}", requireTokenCatalogs);
+    try {
+      for (String catalogName : requireTokenCatalogs) {
+        Optional<HiveConf> hiveConfOpt = buildHiveConf(sparkConf, hadoopConf, catalogName);
+        if (!hiveConfOpt.isPresent()) {
+          continue;
+        }
+
+        HiveConf remoteHmsConf = hiveConfOpt.get();
+        String metastoreUri = sparkConf.get(SPARK_SQL_CATALOG_PREFIX + catalogName + URI_KEY);
+        String principal = sparkConf.get(SPARK_SQL_CATALOG_PREFIX + catalogName + PRINCIPAL_KEY);
+        try {
+          UserGroupInformation currentUser = UserGroupInformation.getCurrentUser();
+          LOG.debug(
+              "Getting Hive delegation token for {} against {} at {}",
+              currentUser.getUserName(),
+              principal,
+              metastoreUri);
+          doAsRealUser(
+              () -> {
+                IMetaStoreClient hmsClient = hmsClientCache.get(metastoreUri);
+                if (hmsClient == null) {
+                  hmsClient = createHmsClient(remoteHmsConf);
+                  hmsClientCache.put(metastoreUri, hmsClient);
+                }
+
+                String tokenStr =
+                    hmsClient.getDelegationToken(currentUser.getUserName(), principal);
+                Token<DelegationTokenIdentifier> hive2Token = new Token<>();
+                hive2Token.decodeFromUrlString(tokenStr);
+                LOG.info("Get Token from hive metastore: {}", hive2Token);
+                creds.addToken(new Text(metastoreUri), hive2Token);
+                return null;
```

Comment:
Since it's returning null that means `obtainDelegationTokens` will never ever been called again unless other tokens are expired. Is this the intention here?